### PR TITLE
Post Firebase

### PR DESCRIPTION
**Pull request by DavidKalajdzic**
Implementation of the whole `PostDataSource` (the interface was defined in #116) for Firebase: `PostFireBaseDataSource` 
+ For the first time a firebase implementation does handle the firebase errors thanks to the Result generic class we defined previously.
+ accessing to unauthorised Posts is producing a negative Result with a very descriptive error message (for example the Post is not public and the user accessing it is not taking part of the event associated to this particular post). Take a look at the interface which explains all those subtleties.
+ lot's of tests!

(my task was in fact to just implement the save of a post #135 but I decided to to it completely so Post functionality is done)

### CODE REVIEW

File: app/src/androidTest/java/com/sdpteam/connectout/post/model/PostTest.java [hunk 1]
```diff
@@ -23,4 +23,23 @@ public void gettersAndSettersTest() {
         assertThat(m.getTitle(), is("title"));
         assertThat(m.getDescription(), is("description"));
     }
+
+    @Test
+    public void equalsTest() {
+        Post m1 = new Post("id", "pid", "eventId", "commentsChatId", new ArrayList<>(), 1, Post.PostVisibility.PUBLIC, "title", "description");
+        Post m2 = new Post("id", "pid", "eventId", "commentsChatId", new ArrayList<>(), 1, Post.PostVisibility.PUBLIC, "title", "description");
+        assertThat(m1.equals(m2), is(true));
```
Comment (MikaelKalajdzic) [inline app/src/androidTest/java/com/sdpteam/connectout/post/model/PostTest.java:31]: what about assertTrue ?

File: app/src/main/java/com/sdpteam/connectout/event/EventFirebaseDataSource.java [hunk 2]
```diff
@@ -110,10 +110,12 @@ public CompletableFuture<Boolean> leaveEvent(String eventId, String participantI
      */
     @Override
     public CompletableFuture<Event> getEvent(String eventId) {
+        System.out.println("Fetching event from firebase: " + eventId);
```
Comment (MikaelKalajdzic) [inline app/src/main/java/com/sdpteam/connectout/event/EventFirebaseDataSource.java:113]: Do we need these logs?

File: app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java [hunk 3]
```diff
@@ -1,63 +1,169 @@
...
+import com.google.firebase.database.DataSnapshot;
+import com.google.firebase.database.FirebaseDatabase;
+import com.sdpteam.connectout.event.Event;
+import com.sdpteam.connectout.event.EventFirebaseDataSource;
+import com.sdpteam.connectout.event.nearbyEvents.filter.EventParticipantIdFilter;
+import com.sdpteam.connectout.event.nearbyEvents.filter.ProfilesFilter;
+import com.sdpteam.connectout.utils.Result;
 
 public class PostFirebaseDataSource implements PostDataSource {
+    public static boolean forceFail = false; // workaround because Mocking the isSuccessful() of the Task class is not possible
```
Comment (MikaelKalajdzic) [inline app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java:22]: static variables in upper case letters

File: app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java [hunk 4]
```diff
@@ -1,63 +1,169 @@
...
+                    post.getDescription());
+        }
+        final Post finalPost = post;
+
+        CompletableFuture<Result<String>> result = new CompletableFuture<>();
+
+        FirebaseDatabase.getInstance().getReference().child(POSTS).child(post.getId()).setValue(post)
+                .addOnCompleteListener(task -> {
+                    if (task.isSuccessful() && !forceFail) {
+                        System.out.println("Post saved successfully");
```
Comment (MikaelKalajdzic) [inline app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java:41]: Useless log

File: app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java [hunk 5]
```diff
@@ -1,63 +1,169 @@
...
 
-        return future;
+            FirebaseDatabase.getInstance().getReference().child(POSTS).get().addOnCompleteListener(task -> {
+                if (task.isSuccessful() && task.getResult().exists() && !forceFail) {
+                    final DataSnapshot dataSnapshot = task.getResult();
+                    final List<Post> postList = StreamSupport.stream(dataSnapshot.getChildren().spliterator(), false)
+                            .map(childSnapshot -> childSnapshot.getValue(Post.class))
+                            .filter(Objects::nonNull)
+                            .filter(post -> {
+                                if (post.getVisibility() != null && post.getVisibility().equals(SEMIPRIVATE) && post.getEventId() != null) {
```
Comment (MikaelKalajdzic) [inline app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java:122]: Can you refactor this condition as a method on post directly and give it a cool name

File: app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java [hunk 6]
```diff
@@ -1,63 +1,169 @@
...
     }
 
-    //TODO fix through firebase task
     @Override
     public CompletableFuture<Result<List<Post>>> fetchPostMadeByUser(String userId, String authorId) {
-        return fetchAllPosts(userId);
+        return fetchAllPostsFiltered(userId, post -> post.getProfileId().equals(authorId));
+    }
+
+    private CompletableFuture<Result<List<Post>>> fetchAllPostsFiltered(String userId, Predicate<Post> postPredicateFilter) {
```
Comment (MikaelKalajdzic) [inline app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java:156]: I would rename postPredicateFilter to postFilter simply

**Review by MikaelKalajdzic — COMMENTED** [review]
Great! good tests
Just a few minor changes, look at the comments

File: app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java (hunk 3 above)
Comment (DavidKalajdzic) [reply app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java:22]: Oupsie

File: app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java (hunk 6 above)
Comment (DavidKalajdzic) [reply app/src/main/java/com/sdpteam/connectout/post/model/PostFirebaseDataSource.java]: Yeah, was struggling finding a clear name ahah

**Review by MikaelKalajdzic — APPROVED** [review]
Perfect, nice job

**Pull request merged**